### PR TITLE
Fix data race in `thread::scope()`

### LINKED ISSUE #98498: Data race in thread::scope
The following code sometimes causes a data race to be reported by Miri:
```rust
#![feature(atomic_from_mut, inline_const)]
use std::sync::atomic::{AtomicBool, Ordering};

fn main() {
let mut some_bools = [const { AtomicBool::new(false) }; 10];

let view: &mut [bool] = AtomicBool::get_mut_slice(&mut some_bools);
assert_eq!(view, [false; 10]);
view[..5].copy_from_slice(&[true; 5]);

std::thread::scope(|s| {
    for t in &some_bools[..5] {
        s.spawn(move || assert_eq!(t.load(Ordering::Relaxed), true));
    }

    for f in &some_bools[5..] {
        s.spawn(move || assert_eq!(f.load(Ordering::Relaxed), false));
    }
});
}
```
Miri reports:
```
note: tracking was triggered
   --> /home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/thread/scoped.rs:133:17
    |
133 |       let scope = Scope {
    |  _________________^
134 | |         data: ScopeData {
135 | |             num_running_threads: AtomicUsize::new(0),
136 | |             main_thread: current(),
...   |
140 | |         scope: PhantomData,
141 | |     };
    | |_____^ created allocation with id 1082
    |
    = note: inside `std::thread::scope::<[closure@atomic.rs:12:20: 20:2], ()>` at /home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/thread/scoped.rs:133:17
note: inside `main` at atomic.rs:12:1
   --> atomic.rs:12:1
    |
12  | / std::thread::scope(|s| {
13  | |     for t in &some_bools[..5] {
14  | |         s.spawn(move || assert_eq!(t.load(Ordering::Relaxed), true));
15  | |     }
...   |
19  | |     }
20  | | });
    | |__^

error: Undefined Behavior: Data race detected between Deallocate on Thread(id = 0, name = "main") and Read on Thread(id = 2) at alloc1082+0x8 (current vector clock = VClock([114, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8]), conflicting timestamp = VClock([113, 0, 9, 9, 9, 0, 0, 0, 0, 0, 9]))
  --> atomic.rs:12:1
   |
12 | / std::thread::scope(|s| {
13 | |     for t in &some_bools[..5] {
14 | |         s.spawn(move || assert_eq!(t.load(Ordering::Relaxed), true));
15 | |     }
...  |
19 | |     }
20 | | });
   | |__^ Data race detected between Deallocate on Thread(id = 0, name = "main") and Read on Thread(id = 2) at alloc1082+0x8 (current vector clock = VClock([114, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8]), conflicting timestamp = VClock([113, 0, 9, 9, 9, 0, 0, 0, 0, 0, 9]))
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information

   = note: inside `main` at atomic.rs:12:1
```
The deallocation occurs when `scope` returns and its local variable `scope` gets deallocated.

The read occurs here (I am fairly sure):

https://github.com/rust-lang/rust/blob/7098a714e8dffe32ae10a27a18a3d73b23431a55/library/std/src/thread/scoped.rs#L58-L59

This reads the contents of the field `self.main_thread` *after* the `fetch_sub`. But the moment the `fetch_sub` is done, the memory backing `self` might be deallocated. That makes the read a potential use-after-free, and a race with the deallocation.

Cc @m-ou-se

### PR DESCRIPTION
**Pull request by danielhenrymantilla**
Fixes #98498, based on @RalfJung's suggestions (taking the non-`Arc` approach; for an `Arc`-based alternative, see #98503).

r? @m-ou-se

### CODE REVIEW

File: library/std/src/thread/scoped.rs [hunk 1]
```diff
@@ -51,12 +54,24 @@ impl ScopeData {
...
+        let main_thread = self.main_thread().clone();
         if self.num_running_threads.fetch_sub(1, Ordering::Release) == 1 {
-            self.main_thread.unpark();
+            // By now, `num_running_threads` is `0`, so when `scope()` in the main thread
+            // is unparked, it will complete its business and deallocate `*self`!
+            // Two things to look after:
+            //   - it can spuriously unpark / wake up, **so `self` can no longer be used**, even
+            //     before we, ourselves, unpark. Hence why we've cloned the `main_thread`'s handle.
+            //   - no matter how it unparks, `*self` may be deallocated before this function
+            //     returns, so all of `*self` data, **including `main_thread`**, must be interiorly
```
Comment (RalfJung) [inline library/std/src/thread/scoped.rs:72]:
```suggestion
            //     returns, so all of `*self` fields, **including `main_thread`**, must be interiorly
```

File: library/std/src/thread/scoped.rs [hunk 2]
```diff
@@ -51,12 +54,24 @@ impl ScopeData {
...
+            //     mutable. See https://github.com/rust-lang/rust/issues/55005 for more info.
```
Comment (RalfJung) [inline library/std/src/thread/scoped.rs:73]: I'd suggest to link to https://github.com/rust-lang/rust/pull/98017, which adjusts the UnsafeCell docs for this.

**Comment by RalfJung** [comment]
See https://github.com/rust-lang/rust/pull/98503 for the `Arc` alternative. :)

This here avoids a pointer indirection, but it also `clone`s once per thread (when the thread finishes). And it needs `unsafe` code. :shrug: not sure what is preferable.^^

I'll test this PR in Miri... once it actually builds:
```
error[E0412]: cannot find type `UnsafeCell` in this scope
  --> /home/r/src/rust/rustc/library/std/src/thread/scoped.rs:44:18
   |
44 |     main_thread: UnsafeCell<Thread>,
   |                  ^^^^^^^^^^ not found in this scope
   |
```
(After fixing this it complains that `UnsafeCell` is non-`Sync`)

**Comment by danielhenrymantilla** [comment]
Yeah, I have to fix these things: I submitted the PR from the web UI 😅 

@rustbot author

**Comment by RalfJung** [comment]
Hm, I am still seeing Stacked Borrows violations with this PR:
```
error: Undefined Behavior: deallocating while item is protected: [SharedReadOnly for <45060> (call 13916)]
  --> atomic.rs:11:1
   |
11 | / std::thread::scope(|s| {
12 | |     for t in &some_bools[..5] {
13 | |         s.spawn(move || assert_eq!(t.load(Ordering::Relaxed), true));
14 | |     }
...  |
18 | |     }
19 | | });
   | |__^ deallocating while item is protected: [SharedReadOnly for <45060> (call 13916)]
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
           
   = note: inside `main` at atomic.rs:11:1

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
```

**Comment by RalfJung** [comment]
Oh, I think what is happening is that the *padding between the fields* is not in an `UnsafeCell` and hence Stacked Borrows does not allow it to be deallocated while a reference points to it... :/

**Comment by danielhenrymantilla** [comment]
Oh no. Ok, at this point, the sketches I have to tackle padding too, make the PR become way more unwieldy[^1], to my taste, than your now definitely simpler `Arc` one. I'm thus gonna close this unless we were to somehow become convinced of some actual benefits from this `Arc`-less approach that could make it worth the code complexity.

[^1]: See https://github.com/rust-lang/rust/pull/98017#discussion_r906825431 for more info

**Pull request closed without merging**